canopycover.py

```diff
@@ -0,0 +1,358 @@
+
+    Return:
+        Returns an array of trait values taken from the traits parameter
+    """
+    # compose the summary traits
+    fields = get_fields()
+    trait_list = []
+    for field_name in fields:
+        if field_name in traits:
+            trait_list.append(traits[field_name])
+        else:
+            trait_list.append(get_default_trait(field_name))
+
+    return trait_list
+
+
+def setup_default_traits(traits: dict, args: argparse.Namespace, full_md: list) -> dict:
+    """Overrides trait values based upon command line parameters and loaded metadata
+    Arguments:
+        traits: the current set of traits
+        args: command line arguments which may override values
+        full_md: the loaded metadata which is checked for default values
+    Return:
+        A copy of the traits with updated values, or the original traits if nothing was changed
+    """
+    new_traits = copy.deepcopy(traits)
+    traits_modified = False
+
+    # Check metadata
+    if full_md:
+        for one_md in full_md:
+            if 'germplasmName' in one_md:
+                new_traits['species'] = one_md['germplasmName']
+                traits_modified = True
+
+    # Check command line parameters
+    if args.germplasmName is not None:
+        new_traits['species'] = args.germplasmName
+        traits_modified = True
+    if args.citationAuthor is not None:
+        new_traits['citation_author'] = args.citationAuthor
+        traits_modified = True
+    if args.citationTitle is not None:
+        new_traits['citation_title'] = args.citationTitle
+        traits_modified = True
+    if args.citationYear is not None:
+        new_traits['citation_year'] = args.citationYear
+        traits_modified = True
+
+    # Return the appropriate traits
+    return new_traits if traits_modified else traits
+
+
+def calculate_canopycover_masked(pxarray: np.ndarray) -> float:
+    """Return greenness percentage of given numpy array of pixels.
+    Arguments:
+      pxarray (numpy array): rgba image where alpha 255=data and alpha 0=NoData
+    Returns:
+      (float): greenness percentage
+    """
+    nonzeros = np.count_nonzero(pxarray)
+    ratio = nonzeros/float(pxarray.size)
+    # Scale ratio from 0-1 to 0-100
+    ratio *= 100.0
+
+    return ratio
+
+    # # If > 75% is NoData, return a -1 ccvalue for omission later
+    # total_size = pxarray.shape[0] * pxarray.shape[1]
+    # nodata = np.count_nonzero(pxarray[:, :, 3] == 0)
+    # nodata_ratio = nodata / float(total_size)
+    # if nodata_ratio > 0.75:
+    #     return -1
+    #
+    # # For masked images, all pixels with rgb>0,0,0 are considered canopy
+    # data = pxarray[pxarray[:, :, 3] == 255]
+    # canopy = len(data[np.sum(data[:, 0:3], 1) > 0])
+    # ratio = canopy / float(total_size - nodata)
+    # # Scale ratio from 0-1 to 0-100
+    # ratio *= 100.0
+    #
+    # return ratio
+
+
+def centroid_as_json(geom: ogr.Geometry) -> str:
```

Comment:
why not put this in the agpypeline?

Reply:
This will also be removed from here when moving to the rgb plot code template template. It can be put into the library